### PR TITLE
Add constraint to restrict IAM bindings

### LINKED ISSUE #21: Add constraint template for limiting IAM roles
For example, forbidding the owner IAM role.

### PR DESCRIPTION
**Pull request by morgante**
This constraint provides feature parity with the [existing Forseti IAM rules](https://github.com/forseti-security/forseti-security/blob/master/rules/iam_rules.yaml).

Fixes #21.

### CODE REVIEW

File: policies/templates/gcp_iam_allowed_bindings.yaml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+apiVersion: templates.gatekeeper.sh/v1alpha1
+kind: ConstraintTemplate
+metadata:
+  name: gcp-iam-allowed-bindings
+spec:
+  crd:
+    spec:
+      names:
+        kind: GCPIAMAllowedBindingsConstraintV1
+        plural: GCPIAMAllowedBindingsConstraintsV1
```
Comment (t12g) [inline policies/templates/gcp_iam_allowed_bindings.yaml]: let's keep this all lowercase and remove the singular field

File: policies/templates/gcp_iam_allowed_bindings.yaml [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        singular: GCPIAMAllowedBindingsConstraintV1
+      validation:
+        openAPIV3Schema:
+          properties:
+            mode:
```
Comment (t12g) [inline policies/templates/gcp_iam_allowed_bindings.yaml]: can we document the parameters?

File: samples/iam_blacklist_role.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+apiVersion: constraints.gatekeeper.sh/v1alpha1
+kind: GCPIAMAllowedBindingsConstraintV1
+metadata:
+  name: blacklist_cloudasset_role
```
Comment (t12g) [inline samples/iam_blacklist_role.yaml]: this name doesn't quite match the description?

File: validator/iam_allowed_bindings.rego [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package templates.gcp.GCPIAMAllowedBindingsConstraintV1
+
+import data.validator.gcp.lib as lib
+
+# If a primary domain is whitelisted, all of its sub domains are whitelisted as well.
```
Comment (t12g) [inline validator/iam_allowed_bindings.rego]: delete this comment?

File: validator/iam_allowed_bindings_test.rego [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	found_violations := find_violations with data.test_constraints as constraints
+
+	violation := found_violations[_]
+}
+
+test_whitelist_role_domain_violations {
+	count(whitelist_role_domain_violations) = 1
+	violation := whitelist_role_domain_violations[_]
+	violation.details.role == "roles/owner"
+}
```
Comment (t12g) [inline validator/iam_allowed_bindings_test.rego]: can we also check the member?

File: validator/iam_allowed_bindings_test.rego [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+	found_violations := find_violations with data.test_constraints as constraints
+
+	violation := found_violations[_]
+}
+
+test_blacklist_public_violations {
+	count(blacklist_public_violations) = 2
+	violation := whitelist_role_domain_violations[_]
+}
```
Comment (t12g) [inline validator/iam_allowed_bindings_test.rego]: can we add positive tests for whitelist & blacklist?

File: validator/iam_allowed_bindings_test.rego [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+	constraints := [fixture_constraints.iam_allowed_bindings_blacklist_role]
+
+	found_violations := find_violations with data.test_constraints as constraints
+
+	violation := found_violations[_]
+}
+
+test_blacklist_role_violations {
+	count(blacklist_role_violations) = 2
+	violation := blacklist_role_violations[_]
```
Comment (t12g) [inline validator/iam_allowed_bindings_test.rego]: curious: do we need this "violation := blacklist_role_violations[_]" line if we're not examining the violation?

File: validator/iam_allowed_bindings_test.rego (hunk 6 above)
Comment (morgante) [reply validator/iam_allowed_bindings_test.rego]: What do you mean by positive tests?

File: validator/iam_allowed_bindings_test.rego (hunk 7 above)
Comment (t12g) [reply validator/iam_allowed_bindings_test.rego]: where no violation is found

Comment (t12g) [reply validator/iam_allowed_bindings_test.rego]: er, entered the comment in the wrong place. Meant to reply to the "positive tests" comment....

**Comment by morgante** [comment]
@t12g Addressed feedback.

**Pull request merged**